s3same/iam.py

```diff
@@ -0,0 +1,88 @@
+import json
+from botocore.exceptions import ClientError
+
+IAMName = 's3same_travis'
+
+def _policy_string(bucket):
+    return json.dumps({
+        "Version": "2012-10-17",
+        "Statement": [
+            {
+                "Action": [
+                    "s3:ListBucket"
+                    ],
+                "Effect": "Allow",
+                "Resource": [
+                    "arn:aws:s3:::{}".format(bucket)
+                    ]
+                },
+            {
+                "Action": [
+                    "s3:PutObject",
+                    "s3:PutObjectAcl"
+                    ],
+                "Effect": "Allow",
+                "Resource": [
+                    "arn:aws:s3:::{}/*".format(bucket)
+                    ]
+                }
+            ]
+        })
+
+def _all_policies(iam, **kwargs):
+    marker = None
+    while True:
+        if marker:
+            kwargs['Marker'] = marker
+        response = iam.list_policies(**kwargs)
+        for policy in response.get('Policies', []):
+            yield policy
+        marker = response.get('Marker')
+        if not response.get('IsTruncated') or marker is None:
+            break
+    
+def _policy_arn(iam, bucket):
+    policy_arn = None
+    try:
+        # Try creating the policy...
+        response = iam.create_policy(
+                PolicyName=IAMName,
+                PolicyDocument=_policy_string(bucket),
+                )
+        # ... and getting its ARN from the response.
+        policy_arn = response.get('Policy', {}).get('Arn')
+    except ClientError as e:
+        if e.response['Error']['Code'] != 'EntityAlreadyExists':
+            raise
+        # If the policy already exists, go find it and get its ARN.
+        policy_arn = next(
+                (p.get('Arn')
+                    for p in _all_policies(iam, Scope='Local')
+                    if p.get('PolicyName') == IAMName
+                    ),
+                None)
```

Comment:
I agree about the name of the `_all_policies` function being more properly named something else if the if condition is included (maybe `_get_existing_policy`). To me, calling `next()` over a generator comprehension to iterate over the yields of a generator function actually reduces readability.

But it could be that we're just approaching this from different directions. In my mind, the method currently called `_policy_arn` (which I would probably call something like `_create_policy`) is coordinating the creation or retrieval of a Policy. It passes execution to the AWS SDK to insert a policy, then if that fails passes execution to `all_policies` to retrieve the existing one.

In that sense, it may be more a philosophical difference, triggered at least partially on my part by the atypical use of `next()` with nested generators.